pennylane/_qubit_device.py

```diff
@@ -161,6 +174,14 @@ def execute(self, circuit, **kwargs):
         Returns:
             array[float]: measured value(s)
         """
+        if self._caching:
+            try:  # TODO: Remove try/except when circuit is always QuantumTape
+                circuit_hash = circuit.graph.hash
```

Comment:
@trbromley and I had a long discussion about this 😆 it really depends how we want to support this on the user level, e.g., do we want equivalent circuits (see below) to have the same hash, or restrict it more strictly to identical circuits?

```python
def circuit1(x, y):
    qml.RX(x, wires=0)
    qml.RY(y, wires=1)
    return qml.expval(qml.PauliZ(0))

def circuit2(x, y):
    qml.RY(y, wires=1)
    qml.RX(x, wires=0)
    return qml.expval(qml.PauliZ(0))
```

Reply:
Maybe a good middle ground is to make it an option; have a boolean argument to decide whether circuits must be identical or equivalent to have the same hash, and use feedback to choose a sensible default (make the DAG construction false by default).

Similar to the gradient methods, where there is a `use_graph=` kwarg.

Reply:
> Do we have a plan to implement hashing of the tape without using the DAG, in a way that is hopefully faster?

Once we have a way of serializing the tape, we can use this to implement hashing. Only tapes that produce identical serializations will produce the same hash. However, it will be less general than circuit graph hashing --- gate order will matter.

> Now that we are on the device side, I'm not sure if it makes sense to just cache the tape parameters.

Agree --- since a device can be used for multiple tapes, we must hash the entire circuit structure.